File: chain-responsibility/java/src/AbstractHandler.java
package src;

// 抽象处理类，所有处理变成链式，可以互动干涉，动态组合
public abstract class AbstractHandler {
   // 形成职责链
   private AbstractHandler next;

   // 创建调用链，传入多个handler，按顺序形成链，返回第一个handler
   public static AbstractHandler link(AbstractHandler first, AbstractHandler... chain) {
      AbstractHandler head = first;
      for (AbstractHandler handler : chain) {
         head.next = handler;
         head = handler;
      }
      return first;
   }

   // 子类需要实现的检查方法
   public abstract boolean check(int uid);

   // 继续下一个检查
   protected boolean checkNext(int uid) {
      if (next == null) {
         return true;
      }
      return next.check(uid);
   }
}

File: chain-responsibility/go/src/AbstractHandler.go
package src

// 事件基础类，聚合所有公共方法，所有处理变成链式，可以互动干涉，动态组合
// go中没有抽象类，因此将统一类型抽取为interface，公共方法单独放在一个类或用独立函数替代
type AbstractHandler struct {
}

// 创建调用链，传入多个handler，按顺序形成链，返回第一个handler
func (a *AbstractHandler) Link(first Handler, chain ...Handler) Handler {
  var head = first
  for _, handler := range chain {
    head.SetNext(handler)
    head = handler
  }

  return first
}

// 继续下一个检查
func (a *AbstractHandler) CheckNext(hander Handler, uid int) bool {
  // 如果没有next了表示链条检查完毕，返回true
  if hander.GetNext() == nil {
    return true
  }
  return hander.GetNext().Check(uid)
}
